### PR TITLE
Statistics

### PR DESCRIPTION
**Pull request by alpha-carinae29**
add a CSV logger that log detected objects (persons) information and violated social distancing incidents with user defined frequency into CSV files.
related to #19

### CODE REVIEW

**Comment by mhejrati** [comment]
@mrn-mln @JsonSadler can you please provide feedback on this PR?

File: applications/smart-distancing/libs/loggers/csv_logger.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+    Args:
+        object_item: it is a dictionary that contains an detected object information after postprocessing
+        frame_number: current frame number
+
+    Returns:
+        A transformed version of object_item to a dictionary with only scalar values. It also contains an item
+        for frame number.
+
+    """
```
Comment (mrn-mln) [inline applications/smart-distancing/libs/loggers/csv_logger.py]: Please capitalize the first letter of each sentence of the docstring.

File: applications/smart-distancing/libs/loggers/csv_logger.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    object_dict = {}
+    object_dict.update({"frame_number": frame_number})
+    for key, value in object_item.items():
```
Comment (mrn-mln) [inline applications/smart-distancing/libs/loggers/csv_logger.py]: I suggest renaming "object_item" to "detected_object" or "detected_objects". Be careful using s after variables name if there is more than one object at the dictionary.

File: applications/smart-distancing/libs/loggers/csv_logger.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    Args:
+        object_item: it is a dictionary that contains an detected object information after postprocessing
+        frame_number: current frame number
+
+    Returns:
+        A transformed version of object_item to a dictionary with only scalar values. It also contains an item
+        for frame number.
+
+    """
+    object_dict = {}
```
Comment (mrn-mln) [inline applications/smart-distancing/libs/loggers/csv_logger.py:24]: For the sake of clearance rename it to object**s**_dict if there is more than one object

File: applications/smart-distancing/libs/loggers/csv_logger.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    object_dict.update({"frame_number": frame_number})
+    for key, value in object_item.items():
+        if isinstance(value, (list, tuple)):
+            for i, item in enumerate(value):
+                if isinstance(item, float):
+                    item = round(item, 4)
+                elif isinstance(item, np.float32):
+                    item = round(float(item), 4)
```
Comment (mrn-mln) [inline applications/smart-distancing/libs/loggers/csv_logger.py]: Add a TODO to this line. 
"TODO: Why some items types are float and others are np.float?"

File: applications/smart-distancing/libs/loggers/csv_logger.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    this logger creates two csv file every day in two different directory, one for logging detected objects
+    and one for logging violated social distancing incidents. the file names are the same as recording date.
+
+    Attributes:
+        config: a ConfigEngine object which store all of the config parameters. access  to any parameter
+        is possible by calling get_section_dict method.
+        log_directory: the parent directory that stores all log file.
+        objects_log_directory: a directory inside the log_directory that stores object log files.
+        distances_log_directory: a directory inside the log_directory that stores violated distances log files.
+
```
Comment (mrn-mln) [inline applications/smart-distancing/libs/loggers/csv_logger.py]: Please capitalize the first letter of each sentence of the docstring.
Change class docstring to this format:
"""
 Description.....

:param pr1: Description ...
:param pr2: Description ...
"""

File: applications/smart-distancing/libs/loggers/csv_logger.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+                if isinstance(item, float):
+                    item = round(item, 4)
+                elif isinstance(item, np.float32):
+                    item = round(float(item), 4)
+
+                object_dict.update({str(key) + "_" + str(i): item})
+        else:
+            if isinstance(value, float):
+                value = round(value, 4)
+            elif isinstance(value, np.float32):
```
Comment (mrn-mln) [inline applications/smart-distancing/libs/loggers/csv_logger.py]: "TODO: Why the type of values are different?"

File: applications/smart-distancing/libs/loggers/csv_logger.py (hunk 4 above)
Comment (mrn-mln) [inline applications/smart-distancing/libs/loggers/csv_logger.py]: I think it does work for both types
item = round(float(item, 4))
And If you want to check the instance type you can write something like this:
if (isinstance(item, float) and isinstance(item, npfloat32)): item = round(float(item, 4))

File: applications/smart-distancing/libs/loggers/csv_logger.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+            os.mkdir(self.distances_log_directory)
+
+    def update(self, frame_number, object_list, distances):
+        """write the object and violated distances information of a frame into log files.
+
+        Args:
+            frame_number: current frame number
+            object_list: a list of dictionary where each dictionary stores information of an object (person) in a frame.
+            distances: a 2-d numpy array that stores distance between each pair of objects.
+        """
```
Comment (mrn-mln) [inline applications/smart-distancing/libs/loggers/csv_logger.py:74]: Please capitalize the first letter of each sentence of the docstring.

File: applications/smart-distancing/libs/loggers/csv_logger.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+                               "distance": distances[indices[0], indices[1]]} for indices in violating_objects])
+
+    def extract_violating_objects(self, distances):
+        """extract pair of objects that are closer than the distance threshold.
+
+        Args:
+            distances: a 2-d numpy array that stores distance between each pair of objects.
+
+        Returns:
+            A 2-d numpy array where each row is the ids of the objects that violated the social distancing.
```
Comment (mrn-mln) [inline applications/smart-distancing/libs/loggers/csv_logger.py]: Returns:
        violating_objects: A 2-d numpy ...

File: applications/smart-distancing/libs/loggers/loggers.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+        name: logger name, at this time only csv_logger is supported. you can implement your own logger
+        by following csv_logger implementation as an example.
+        logger: Logger object which is built based on the name parameter specified in config file.
+        time_interval: specifies how often the logger should log information. For example with time_interval of 0.5
+        the logger log the information every 0.5 seconds. (set zero to log in every frame)
+        frame_number: keep track of current frame number of the video. This will be used by update method to decide
+        whether its time to log information or logging should be skipped for this frame (based on logger time interval.)
+
+
+    """
```
Comment (mrn-mln) [inline applications/smart-distancing/libs/loggers/loggers.py:19]: Using the mentioned format for docstring

**Review by mrn-mln — CHANGES_REQUESTED** [review]
@alpha-carinae29 Thank you. I reviewed your PR and left some comments.

File: applications/smart-distancing/libs/loggers/csv_logger.py (hunk 3 above)
Comment (alpha-carinae29) [reply applications/smart-distancing/libs/loggers/csv_logger.py:24]: no it only contains one object information

File: applications/smart-distancing/libs/loggers/csv_logger.py (hunk 5 above)
Comment (alpha-carinae29) [reply applications/smart-distancing/libs/loggers/csv_logger.py]: I change the docstring format to be consistent with the rest of the code base but I think it is better to choose the Google's style later.

**Review by mhejrati — APPROVED** [review]
Looks good, thanks @alpha-carinae29

**Comment by mhejrati** [comment]
@alpha-carinae29 can you please resolve the conflicts and resubmit.

**Comment by alpha-carinae29** [comment]
@mhejrati I resolved conflicts

**Pull request merged**